sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -685,13 +163,106 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
   }
 
   /**
+   * Returns a function to generate equal expression in Java
+   */
+  def equalFunc(dataType: DataType): ((Term, Term) => Code) = dataType match {
+    case BinaryType => { case (eval1, eval2) => s"java.util.Arrays.equals($eval1, $eval2)" }
+    case dt if isNativeType(dt) => { case (eval1, eval2) => s"$eval1 == $eval2" }
```

Comment:
Yea but the definition of native type (4 lines down) is : " List of data types that have special accessors and setters in [[Row]]." Which means in the future we might break this.

I'd make it explicit with primitive type here.